File: src/lib/wow/db/caches/db-model-file-data.ts
/**
 * ModelFileData.db2 cache, ported from wow.export (src/js/db/caches/DBModelFileData.js).
 */
import { write } from '@/lib/wow/log';

import { doOnce } from '../../formats/generics';
import { WDCReader } from '../wdc-reader';

const modelResIDToFileDataID = new Map<number, number[]>();
const fileDataIDs = new Set<number>();

/** Initialize model file data from ModelFileData.db2 */
export const initializeModelFileData = doOnce('initializeModelFileData', async () => {
  if (modelResIDToFileDataID.size > 0) return;
  write('Loading model mapping...');
  const modelFileData = new WDCReader('DBFilesClient/ModelFileData.db2');
  await modelFileData.parse();

  // Using the texture mapping, map all model fileDataIDs to used textures.
  for (const [modelFileDataID, modelFileDataRow] of modelFileData.getAllRows()) {
    // Keep a list of all FIDs for listfile unknowns.
    fileDataIDs.add(modelFileDataID);

    const modelResourcesID = modelFileDataRow.ModelResourcesID as number;
    if (modelResIDToFileDataID.has(modelResourcesID)) modelResIDToFileDataID.get(modelResourcesID)!.push(modelFileDataID);
    else modelResIDToFileDataID.set(modelResourcesID, [modelFileDataID]);
  }
  write('Loaded model mapping for %d models', modelResIDToFileDataID.size);
});

/** Retrieve model file data IDs by a model resource ID. */
export function getModelFileDataID(modelResID: number): number[] | undefined {
  return modelResIDToFileDataID.get(modelResID);
}

/** Retrieve a list of all file data IDs cached from ModelFileData.db2 */
export function getFileDataIDs(): Set<number> {
  return fileDataIDs;
}

export function getModelFileDataCacheStats(): { modelResIDs: number; fileDataIDs: number } {
  return { modelResIDs: modelResIDToFileDataID.size, fileDataIDs: fileDataIDs.size };
}

export function resetModelFileDataCache(): void {
  modelResIDToFileDataID.clear();
  fileDataIDs.clear();
}

export default { initializeModelFileData, getModelFileDataID, getFileDataIDs };

File: golang/internal/wow/db/caches/db_model_file_data.go
package caches

import (
	"context"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/wow/db"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var (
	modelResIDToFileDataID = make(map[uint32][]uint32)
	modelFileDataIDs       = make(map[uint32]struct{})
	modelFileDataOnce      sync.Once
	modelFileDataErr       error
)

// InitializeModelFileData loads ModelFileData.db2.
func InitializeModelFileData(ctx context.Context) error {
	modelFileDataOnce.Do(func() {
		if len(modelResIDToFileDataID) > 0 {
			return
		}
		log.Write("Loading model mapping...")
		reader := db.NewWDCReader("DBFilesClient/ModelFileData.db2", nil)
		if err := reader.Parse(ctx, nil); err != nil {
			modelFileDataErr = err
			return
		}
		for modelFileDataID, row := range reader.GetAllRows() {
			modelFileDataIDs[modelFileDataID] = struct{}{}
			modelResourcesID := toUint32(row["ModelResourcesID"])
			modelResIDToFileDataID[modelResourcesID] = append(modelResIDToFileDataID[modelResourcesID], modelFileDataID)
		}
		log.Write("Loaded model mapping for %d models", len(modelResIDToFileDataID))
	})
	return modelFileDataErr
}

// GetModelFileDataID returns file data IDs for a model resource ID.
func GetModelFileDataID(modelResID uint32) []uint32 {
	return modelResIDToFileDataID[modelResID]
}

// GetModelFileDataCacheStats returns cache stats.
func GetModelFileDataCacheStats() (modelResIDs, fileDataIDs int) {
	return len(modelResIDToFileDataID), len(modelFileDataIDs)
}

// ResetModelFileDataCache clears the model file data cache.
func ResetModelFileDataCache() {
	modelResIDToFileDataID = make(map[uint32][]uint32)
	modelFileDataIDs = make(map[uint32]struct{})
	modelFileDataOnce = sync.Once{}
	modelFileDataErr = nil
}

func toUint32(v any) uint32 {
	switch x := v.(type) {
	case uint32:
		return x
	case uint8:
		return uint32(x)
	case uint16:
		return uint32(x)
	case uint64:
		return uint32(x)
	case int:
		return uint32(x)
	case int8:
		return uint32(x)
	case int16:
		return uint32(x)
	case int32:
		return uint32(x)
	case int64:
		return uint32(x)
	case float32:
		return uint32(x)
	case float64:
		return uint32(x)
	case []uint32:
		if len(x) > 0 {
			return x[0]
		}
	case []int64:
		if len(x) > 0 {
			return uint32(x[0])
		}
	}
	return 0
}
